self_hosted/main.oomph

```diff
@@ -1,117 +1,278 @@
 import "<stdlib>/io.oomph" as io
+import "<stdlib>/path.oomph" as path
 import "<stdlib>/process.oomph" as process
+import "ast.oomph" as ast
 import "ast2ir.oomph" as ast2ir
 import "ast_transformer.oomph" as ast_transformer
 import "c_output.oomph" as c_output
 import "ir.oomph" as ir
 import "parser.oomph" as parser
-import "tokenizer.oomph" as tokenizer
 
-# TODO: respect quoting rules
+
+class CompilationUnit(
+    Str source_path,
+    List[ast::ToplevelDeclaration] ast,
+):
+    # Returns source paths
+    meth get_dependencies() -> List[Str]:
+        let result = []
+        if self.source_path != "builtins.oomph":
+            result.push("builtins.oomph")
+        foreach top_decl of self.ast:
+            switch top_decl:
+                case ast::Import imp:
+                    result.push(imp.path)
+                case *:
+                    pass
+        return result
+
+    meth create_c_code(c_output::Session session):
+        let ir = ast2ir::convert_program(session.builtins, self.ast, self.source_path, session.symbols)
+        session.create_c_code(ir, self.source_path)
+
+
+# split() and join() don't cover all corner cases
 func split(Str s) -> List[Str]:
+    # TODO: use quoting rules
     let result = s.split(" ")
     while "" in result:
         result.delete_first("")
     return result
 
-func get_c_compiler_command(List[Str] c_paths, Str exepath) -> List[Str]:
-    # TODO: dict
-    let cc = ""
-    let cflags = ""
-    let ldflags = ""
+func join(List[Str] args) -> Str:
+    let result = []
+    foreach arg of args:
+        if " " in arg:
+            result.push("'" + arg + "'")
+        else:
+            result.push(arg)
+    return result.join(" ")
 
-    foreach line of io::read_file("obj/compile_info.txt").split("\n"):
-        if line == "":
-            continue
+
+class Command(List[Str] command, Str human_readable)
+
+func get_c_compiler_command(List[Str] c_paths, Str exepath) -> Command:
+    let cc = new (Str | null)(null)
+    let cflags = (new Str | null)(null)
+    let ldflags = (new Str | null)(null)
```

Comment:
WTF? this works?!!

Reply:
Just understood how it works: `Str | null` is a type, `new Str | null` is a constructor function that gets called. This happens regardless of which way to write parentheses is used. In fact, even this would work: `new Str | null (null)`